### ISSUE TITLE
add APPNAME when starting a SAS9.4 IOM connection

### ISSUE THREAD

**Issue by brunomueller** [sassoftware/vscode-sas-extension#762]
**Is your feature request related to a problem? Please describe.**
Looking at the log file of the Object Spawner we see the following:
2024-01-17T14:29:12,910 INFO  [00000105] :sasdemo - New client connection (20) accepted from server port 8591 for user sasdemo. ... :56408 for APPNAME=SAS Data Integration Studio.
2024-01-17T14:33:34,293 INFO  [00000130] :sasdemo - New client connection (22) accepted from server port 8591 for user sasdemo. ... :56800.
2024-01-17T14:38:32,804 INFO  [00000173] :sasdemo - New client connection (26) accepted from server port 8591 for user sasdemo. ... :61221 for APPNAME=SAS Enterprise Guide.

The first line is the request to start a workspace server from SAS Data Integration Studio
The second line is the request to start a workspace server from the SAS VSCode extension
The third line is the request to start a workspace server from SAS Enterprise Guide

**Describe the solution you'd like**
Would be great to see an APPNAME= as well when starting a workspace server from VSCode

This will help anyone analysing Object Spawner log files

**Environment**
SAS version: SAS9.4 M8
SAS Extension for VSCode: 1.6.0

**Comment by DominiqueLatour** [comment]
I am currently working with Ralph Abbey and Xin Hunt on PROC REGISTERMODEL to register models to Model Manager.  That will enable users to register models from workbench and SAS Studio.  One of the requirements is to have the environment the model was created.  For example, when submitting code form SAS Studio, the following macros are initialized by SAS Studio:  
%let _clientapp = %nrquote(%nrstr(SAS Studio));
%let _clientappabbrev = %nrquote(%nrstr(Studio));
%let _clientappversion=2020;

This enables us to fetch that information.  It would be great that in the VS Code plug for workbench to have similar macros to enable use to retrieve that we are in "VS Code SAS" and have the version.

Casey Smith told me that Bruno had already entered a similar request although not specific to workbench

**Comment by snlwih** [comment]
@DominiqueLatour, sounds to me that what you describe is slighly different from what Bruno is asking for.

Bruno is just asking for providing client info on the request that initiates a server connection, so the server log shows which client connected to it. When I interpret your suggestions correctly, you are asking for client-generated marcro variables that represent the name of the client application and it's version number, so you can take advantage of that in the SAS code you are submitting. 

Could you please create a dedicated GitHub Issue for this, so we can track this easier going forward? Once you have created it, we will add a link to it in this comment thread.

**Comment by snlwih** [comment]
@brunomueller, am I correct in saying that SAS applications like SAS Data Integration Studio and SAS Enterprise Guide only provide their appname and no version information on the request? Just double-checking.

**Comment by brunomueller** [comment]
@snlwih yes the APPNAME= only contains the application name, see the very first entry.

**Comment by snlwih** [comment]
#1264 will be used to track the request for having the SAS extension generate connection type agnostic SAS macro variables that represent things like -name-of-file-being-submitted, name-of-directory-containing-file-being-submitted, client-app-version, client-app-name, ...

**Comment by clangsmith** [comment]
The Powershell code (in client\src\connection\itc\script.ts) we use for ITCLIENT (SAS 9 - remote IOM) connections currently only uses the COM objects directly. It does not load or use the .NET COM interop assemblies (such as SASInterop.dll, SASOManInterop.dll, etc) and types. (EG and AMO do load and use the .NET COM interop assemblies and types extensively.) Using the .NET COM interop assemblies/types makes working with COM *much* easier (than working with the COM objects directly). You can also easily open these .NET COM interop assemblies in a disassembler (such as ILSpy.exe) and see the type details (methods, parameters, properties, etc).

In fact, I have been unable to find a way to get or set the ApplicationName property on the ObjectFactory COM object directly. I have *only* been able to get/set the ApplicationName property using the .NET COM interop types. See the attached setAppNameViaIOM.txt Powershell script (change the extension from ".txt" to ".ps1" to use) for the examples I tried.

Therefore, to implement this feature, I think we will have to switch to using the .NET COM interop assemblies. However, I think we were going to need to switch to the .NET COM interop types anyway to do more advanced IOM related features, such as the V9 file system support. The other big benefit this provides is that we can follow similar patterns of use that EG and AMO have already established.

One small additional challenge this presents though is the need to load these interop assemblies, so our Powershell script can find/load/use the types. I put an example in the attached script that gets the ITCLIENT x86 installation directory from a Windows registry key, thus is able to dynamically load the SAS COM interop assemblies, which live in the ITCLIENT (and EG and AMO) installation directory.

**Comment by clangsmith** [comment]
Here is the Powershell script example I mentioned in previous comment:
[setAppNameViaIOM.txt](https://github.com/user-attachments/files/18497331/setAppNameViaIOM.txt)

**Comment by clangsmith** [comment]
I'm also attaching here the SASOMan.IDL file (with extension changed to ".txt" to allow attaching), which was exported from the OLE-COM Object Viewer for the SASOMan.dll in the ITCLIENT installation directory. It clearly shows ObjectFactoryMulti2 implements the  IApplicationNameContext interface, but again, I don't know enough about COM to know why I can't access the ApplicationName property directly from the ObjectFactoryMulti2 COM object. Regardless, we can accomplish with the SAS .NET COM interop assemblies, and again, that switch seems like the right thing to do.

[SASOMan.txt](https://github.com/user-attachments/files/18497384/SASOMan.txt)

.
